prov/efa: Emulated Atomic Support for Cuda Devices

a-szegel opened the pull request:
Changes to prov/efa to add emulated atomic support for CUDA devices:

1. Added result + fetch memory descriptors to rxr_atomic_ex
2. Switched every memcpy to be ofi_hmem memcpy
3. On the target side, added ofi_hmem memcpy's around emulated atomic impl functions, and performed them on host memory.

Testing:
```
fi_rdm_atomic -D cuda -p efa -E  # passes, and exits cleanly
fi_rdm_atomic -p efa -E                # passes, and exits cleanly
```

Tested both sides no cuda, both sides cuda, and one side cuda, one side host.  There is a big slow down when using atomics with CUDA devices (which is expected).

a-szegel committed:
include/ofi_hmem.h: Make ofi_hmem copies public

Move implementation of ofi_copy_to_hmem() and ofi_copy_from_hmem() from
source to header.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

a-szegel committed:
src/hmem_cuda.c: Add const to param to remove warning

Added const to data parameter of cuPointerSetAttribute function pointer to remove compiler
warning.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

a-szegel committed:
include/ofi_hmem.h: Add function cuda_flush_rdma_operations

Adds the function cuda_flush_rdma_operations() for GDR Copy memory synchronization.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

lrbison commented on prov/efa/src/rxr/rxr_pkt_type_req.c:
```diff
@@ -2280,9 +2292,16 @@ ssize_t rxr_pkt_init_compare_rta(struct rxr_ep *ep, struct rxr_op_entry *tx_entr
-	data_size = ofi_copy_from_iov(data, ep->mtu_size - pkt_entry->pkt_size,
-				      tx_entry->atomic_ex.comp_iov,
-				      tx_entry->atomic_ex.comp_iov_count, 0);
+
+    hmem_iface = ((struct efa_mr*) tx_entry->atomic_ex.compare_desc)->peer.iface;
+    device = ((struct efa_mr*) tx_entry->atomic_ex.compare_desc)->peer.device.reserved;
+
+	data_size = ofi_copy_from_hmem_iov(data, ep->mtu_size - pkt_entry->pkt_size, hmem_iface, device,
```
I think the formatting is off here

a-szegel replied:
ty fixed.

wzamazon commented on prov/efa/src/rxr/rxr_op_entry.h:
```diff
@@ -80,6 +80,8 @@ struct rxr_atomic_ex {
 	int resp_iov_count;
 	struct iovec comp_iov[RXR_IOV_LIMIT];
 	int comp_iov_count;
+	void *result_desc;
```
should be an array of `void *` to support multiple iov

wzamazon commented on prov/efa/src/rxr/rxr_atomic.c:
```diff
@@ -383,6 +380,8 @@ rxr_atomic_readwritemsg(struct fid_ep *ep,
 
 	ofi_ioc_to_iov(resultv, atomic_ex.resp_iov, result_count, datatype_size);
 	atomic_ex.resp_iov_count = result_count;
+	atomic_ex.result_desc = *result_desc;
```
`result_desc` is an array, and can have multiple entry. number of entry is in `recult_ioc`.

wzamazon commented on prov/efa/src/rxr/rxr_op_entry.h:
```diff
@@ -166,7 +168,7 @@ struct rxr_op_entry {
 	struct dlist_entry pending_recv_entry;
 #endif
 
-	size_t efa_outstanding_tx_ops; 
```
This format change is nice, but should not be in this commit

a-szegel replied:
done

wzamazon commented on prov/efa/src/rxr/rxr_pkt_type_misc.c:
```diff
@@ -209,10 +209,10 @@ ssize_t rxr_pkt_init_cts(struct rxr_ep *ep,
+	 * CTS is sent using tx_entry in the emulated longcts read
+	 * protocol. The emulated longcts write and the longcts
 	 * message protocols sends CTS using rx_entry.
-	 * This check ensures appropriate tx_id and rx_id are 
```
Format change should be in a different commit

a-szegel replied:
done

wzamazon commented on prov/efa/src/rxr/rxr_pkt_type_misc.c:
```diff
@@ -637,15 +637,23 @@ void rxr_pkt_handle_atomrsp_recv(struct rxr_ep *ep,
-			0, atomrsp_pkt->data,
-			atomrsp_hdr->seg_length);
+    hmem_iface = ((struct efa_mr*) tx_entry->atomic_ex.result_desc)->peer.iface;
+    device = ((struct efa_mr*) tx_entry->atomic_ex.result_desc)->peer.device.reserved;
```
identation wrong

a-szegel replied:
done

lrbison commented on prov/efa/src/rxr/rxr_pkt_type_req.c:
```diff
@@ -2417,13 +2489,50 @@ int rxr_pkt_proc_dc_write_rta(struct rxr_ep *ep,
+	uint64_t device = efa_mr->peer.device.reserved;
+	int ret;
+
+	/* Step 1: Copy data from device to temporary host buffer */
+	host_data = malloc(iov->iov_len);
+	if (!host_data) {
+		return FI_ENOSPC;
+	}
```
future optimization: use some sort of memory pool so we don't continually alloc/free host memory for this.

darrylabbate replied:
I'm guessing `iov->iov_len` is dynamic (and we want to avoid VLAs). Otherwise it makes more sense to just have the string buffer be stack-allocated. E.g.
```c
char host_data[iov->iov_len];
```

a-szegel replied:
good idea.. switched to VLA's.

darrylabbate commented on prov/efa/src/rxr/rxr_pkt_type_req.c:
```diff
@@ -2417,13 +2489,50 @@ int rxr_pkt_proc_dc_write_rta(struct rxr_ep *ep,
+
+	/* Step 1: Copy data from device to temporary host buffer */
+	host_data = malloc(iov->iov_len);
+	if (!host_data) {
```
Probably worth it to wrap all the error checks in `OFI_UNLIKELY()`

a-szegel replied:
good idea.

darrylabbate commented on prov/efa/src/rxr/rxr_pkt_type_req.c:
```diff
@@ -2417,13 +2489,50 @@ int rxr_pkt_proc_dc_write_rta(struct rxr_ep *ep,
+
+	ret = ofi_copy_from_hmem(FI_HMEM_CUDA, device, host_data, iov->iov_base, iov->iov_len);
+	cuda_flush_rdma_operations();
+	if (!ret) {
+		return ret;
+	}
```
Shouldn't `host_data` be `free`'d before returning?

a-szegel replied:
no longer malloc'ed

darrylabbate commented on prov/efa/src/rxr/rxr_pkt_type_req.c:
```diff
@@ -2417,13 +2489,50 @@ int rxr_pkt_proc_dc_write_rta(struct rxr_ep *ep,
+	/* Step 3: Copy data from host buffer to device */
+	ret = ofi_copy_to_hmem(FI_HMEM_CUDA, device, iov->iov_base, host_data, iov->iov_len);
+	cuda_flush_rdma_operations();
+	if (!ret) {
+		return ret;
+	}
```
Same here (`free` `host_data`?). Maybe turn these into a `goto` where `host_data` is `free`'d

a-szegel replied:
no longer malloc'ed

a-szegel committed:
prov/efa: RXR format fixes

Format fixes for prov/efa/src/rxr/rxr_pkt_type_misc.c and
prov/efa/src/rxr/rxr_pkt_type_req.c.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

a-szegel committed:
prov/efa: HMEM_CUDA support for emulated atomics

Adds support for emulated atomic calls to EFA for Cuda accelerators.
Atomics are only supported on machines that have GDR + RMA support.
This aligns with our overly restrictive FI_HMEM definition.
The first major change was switching every memcpy to be a ofi_hmem_*
memcpy. For fetch and compare operations, I added the result and
compare memory descriptors to rxr_atomic_ex. Lastly, I wrapped
the atomic operations on the target in ofi_hmem_memcpys with a temporary
bounce buffer in which the cpu only atomic operations are performed.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

wzamazon commented on prov/efa/src/rxr/rxr_pkt_type_req.c:
```diff
@@ -2227,8 +2229,14 @@ ssize_t rxr_pkt_init_rta(struct rxr_ep *ep, struct rxr_op_entry *tx_entry,
 	data = (char *)pkt_entry->pkt + hdr_size;
-	data_size = ofi_copy_from_iov(data, ep->mtu_size - hdr_size,
-				      tx_entry->iov, tx_entry->iov_count, 0);
+	hmem_iface = ((struct efa_mr*) tx_entry->desc[0])->peer.iface;
```
`tx_entry->desc[0]` can be NULL for system memory

a-szegel replied:
It will never be NULL.  Atomics are for one sided communication, and in order to use 1 sided communication, you must register all your memory.

a-szegel replied:
The compare, result, and buff must all be registered no-matter the memory type in order to use libfabric atomics AFAIK... I think an assert would be better here if you are concerned.
